### PR TITLE
Allow FileWatcher watch directories to be configured (#608)

### PR DESCRIPTION
**Pull request by mathewc**
In addition to addressing the node_modules issue reported in https://github.com/Azure/azure-webjobs-sdk-script/issues/608, I generalize the file watching to allow users to include their own Shared source directories. Currently we tell people how to share code by putting it in a Shared directory in the Functions root, but we don't watch such directories for changes. Since such directories are user defined, I'm allowing them to be configured like so:

```javascript
{
    "id": "5a709861cab44e68bfed5d2c2fe7fc0c",
    "watchDirectories": [ "Shared" ],
    "sendGrid": {
        "from": "Azure Functions <samples@functions.com>"
    }
}
```

### CODE REVIEW

File: sample/QueueTrigger-CSharp/run.csx [hunk 1]
```diff
@@ -8,10 +10,4 @@ public static void Run(Message message, out string result, TraceWriter log)
     log.Info($"C# Queue trigger function processed message: {message.Id}");
 
     result = message.Value;
-}
-
-public class Message
```
Comment (mathewc) [inline sample/QueueTrigger-CSharp/run.csx]: Before these changes, any changes made to Shared\Message.csx would not be picked up (the host wouldn't restart). Now it does.

File: src/WebJobs.Script/Host/ScriptHost.cs [hunk 2]
```diff
@@ -732,10 +732,26 @@ internal static void ApplyConfiguration(JObject config, ScriptHostConfiguration
                 throw new InvalidOperationException("An 'id' must be specified in the host configuration.");
             }
 
-            JToken watchFiles = (JToken)config["watchFiles"];
```
Comment (mathewc) [inline src/WebJobs.Script/Host/ScriptHost.cs]: Renaming the existing watchFiles to fileWatchingEnabled (similar to fileLoggingEnabled)

File: src/WebJobs.Script/Host/ScriptHost.cs [hunk 3]
```diff
@@ -732,10 +732,26 @@ internal static void ApplyConfiguration(JObject config, ScriptHostConfiguration
...
+                scriptConfig.FileWatchingEnabled = (bool)fileWatchingEnabled;
+            }
+
+            // Configure the set of watched directories, adding the standard built in
+            // set to any the user may have specified
+            if (scriptConfig.WatchDirectories == null)
+            {
+                scriptConfig.WatchDirectories = new Collection<string>();
+            }
+            scriptConfig.WatchDirectories.Add("node_modules");
```
Comment (mathewc) [inline src/WebJobs.Script/Host/ScriptHost.cs]: I think node_modules is the only built in we should include? Any others?

File: sample/QueueTrigger-CSharp/run.csx [hunk 4]
```diff
@@ -1,3 +1,5 @@
+#load "..\Shared\Message.csx"
```
Comment (mathewc) [inline sample/QueueTrigger-CSharp/run.csx]: Is there an F# equivalent of this?

Comment (fabiocav) [inline sample/QueueTrigger-CSharp/run.csx]: F# uses `#load` directives as well.

Comment (fabiocav) [inline sample/QueueTrigger-CSharp/run.csx]: https://github.com/fabiocav/azure-webjobs-sdk-script/blob/12df0cd64146193c17b53cda4f08522345bba9f6/test/WebJobs.Script.Tests/TestScripts/FSharp/LoadScriptReference/run.fsx#L7

File: src/WebJobs.Script/Host/ScriptHost.cs [hunk 5]
```diff
@@ -732,10 +732,26 @@ internal static void ApplyConfiguration(JObject config, ScriptHostConfiguration
...
+            JToken watchDirectories = config["watchDirectories"];
+            if (watchDirectories != null && watchDirectories.Type == JTokenType.Array)
```
Comment (davidebbo) [inline src/WebJobs.Script/Host/ScriptHost.cs]: Seems a bit wrong to just ignore it if it's not an array, instead of hard failing.

File: src/WebJobs.Script/Host/ScriptHost.cs [hunk 6]
```diff
@@ -903,6 +922,18 @@ private void OnFileChanged(object sender, FileSystemEventArgs e)
             }
         }
 
+        internal static string GetRelativeDirectory(string path, string scriptRoot)
+        {
+            string directory = path.Substring(scriptRoot.Length).TrimStart(Path.DirectorySeparatorChar);
```
Comment (davidebbo) [inline src/WebJobs.Script/Host/ScriptHost.cs]: Might be a good sanity to check that path in fact starts with scriptRoot.

File: src/WebJobs.Script/Config/ScriptHostConfiguration.cs [hunk 7]
```diff
@@ -49,6 +49,14 @@ public ScriptHostConfiguration()
         public bool FileWatchingEnabled { get; set; }
 
         /// <summary>
+        /// Gets the or sets the collection of directories (relative to RootScriptPath) that
+        /// should be monitored for changes. If FileWatchingEnabled is true, these directories
+        /// will be monitored. When a file is added/modified/deleted in any of these
+        /// directories, the host will restart.
+        /// </summary>
+        public Collection<string> WatchDirectories { get; set; }
```
Comment (davidebbo) [inline src/WebJobs.Script/Config/ScriptHostConfiguration.cs]: Use ICollection<string>

File: src/WebJobs.Script/Host/ScriptHost.cs [hunk 8]
```diff
@@ -732,10 +732,26 @@ internal static void ApplyConfiguration(JObject config, ScriptHostConfiguration
...
             {
-                scriptConfig.FileWatchingEnabled = (bool)watchFiles;
+                scriptConfig.FileWatchingEnabled = (bool)fileWatchingEnabled;
+            }
+
+            // Configure the set of watched directories, adding the standard built in
+            // set to any the user may have specified
+            if (scriptConfig.WatchDirectories == null)
+            {
+                scriptConfig.WatchDirectories = new Collection<string>();
```
Comment (davidebbo) [inline src/WebJobs.Script/Host/ScriptHost.cs]: `HashSet<string>` feels more correct here.

File: src/WebJobs.Script/Host/ScriptHost.cs [hunk 9]
```diff
@@ -891,9 +907,12 @@ private void NotifyInvoker(string functionName, Exception ex)
...
-            string fileName = Path.GetFileName(e.Name);
+            string directory = GetRelativeDirectory(e.FullPath, ScriptConfig.RootScriptPath);
+            bool isWatchedDirectory = ScriptConfig.WatchDirectories.Contains(directory, StringComparer.OrdinalIgnoreCase);
 
-            if (((string.Compare(fileName, ScriptConstants.HostMetadataFileName, StringComparison.OrdinalIgnoreCase) == 0) ||
+            string fileName = Path.GetFileName(e.Name);
+            if (isWatchedDirectory ||
+                ((string.Compare(fileName, ScriptConstants.HostMetadataFileName, StringComparison.OrdinalIgnoreCase) == 0) ||
                 string.Compare(fileName, ScriptConstants.FunctionMetadataFileName, StringComparison.OrdinalIgnoreCase) == 0) ||
                 (Directory.EnumerateDirectories(ScriptConfig.RootScriptPath).Count() != _directoryCountSnapshot))
```
Comment (davidebbo) [inline src/WebJobs.Script/Host/ScriptHost.cs]: This conditional is a bit hard to read through and could use a comment. e.g. explicitely list in plain English the conditions that trigger a restart. I'd also yank a bunch of extra parenthesis which makes it look more complicated that it probably is :)

File: src/WebJobs.Script/Host/ScriptHost.cs (hunk 8 above)
Comment (mathewc) [inline src/WebJobs.Script/Host/ScriptHost.cs]: Done

File: src/WebJobs.Script/Host/ScriptHost.cs (hunk 5 above)
Comment (mathewc) [inline src/WebJobs.Script/Host/ScriptHost.cs]: Logged https://github.com/Azure/azure-webjobs-sdk-script/issues/614 to take care of this across the board. It will take a bit of time.

File: src/WebJobs.Script/Host/ScriptHost.cs (hunk 6 above)
Comment (mathewc) [inline src/WebJobs.Script/Host/ScriptHost.cs]: Done

File: src/WebJobs.Script/Host/ScriptHost.cs [hunk 10]
```diff
@@ -732,10 +732,26 @@ internal static void ApplyConfiguration(JObject config, ScriptHostConfiguration
...
             {
-                scriptConfig.FileWatchingEnabled = (bool)watchFiles;
+                scriptConfig.FileWatchingEnabled = (bool)fileWatchingEnabled;
+            }
+
+            // Configure the set of watched directories, adding the standard built in
+            // set to any the user may have specified
+            if (scriptConfig.WatchDirectories == null)
+            {
+                scriptConfig.WatchDirectories = new HashSet<string>();
```
Comment (davidebbo) [inline src/WebJobs.Script/Host/ScriptHost.cs]: nit: I'd specify `StringComparer.OrdinalIgnoreCase` in the ctor instead of the `.Contains` call below.

File: src/WebJobs.Script/Host/ScriptHost.cs (hunk 9 above)
Comment (mathewc) [inline src/WebJobs.Script/Host/ScriptHost.cs]: Done

File: src/WebJobs.Script/Host/ScriptHost.cs (hunk 10 above)
Comment (mathewc) [inline src/WebJobs.Script/Host/ScriptHost.cs]: Done

**Comment by davidebbo** [comment]
:shipit:
